### ISSUE TITLE
Add most-disruptive-allowed-action support

### ISSUE THREAD

**Issue by raddaoui** [hashicorp/terraform-provider-google#5654]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

most-disruptive-allowed-action (https://cloud.google.com/compute/docs/instance-groups/rolling-out-updates-to-managed-instance-groups#updating_selected_instances) is missing from the Update policy in `google_compute_region_instance_group_manager`

When running stateful resources with ephemeral disks in a managed instance group, this field helps against data loss.

### New or Affected Resource(s)

* google_compute_region_instance_group_manager

### Potential Terraform Configuration

update_policy {
  minimal_action               = "REFRESH"
  most_disruptive_allowed_action   = "RESTART"
}

**Comment by mgaidia** [comment]
Other affected resource:
- google_compute_instance_group_manager

**Comment by ahsan518** [comment]
👍

**Comment by ndmckinley** [comment]
Hi, could you elaborate a little bit about what you'd like to do with this field?  You're saying you'd like to prevent the instances from being recreated, is that right?

**Comment by raddaoui** [comment]
Hi @ndmckinley, that's right we want to prevent instances from being recreated, you can specify both minimal-action and most-disruptive-allowed-action  in the managed instance group with the Google API as you can see [here ](https://cloud.google.com/compute/docs/instance-groups/rolling-out-updates-to-managed-instance-groups#updating_selected_instances) but in terraform only the [minimal-action](https://www.terraform.io/docs/providers/google/r/compute_instance_group_manager.html#minimal_action) option is exposed. we use terraform to provision resources so we want to be able to specify this from terraform as well.

**Comment by ndmckinley** [comment]
Hi!

Okay, we have investigated, and this is fairly nuanced, so I want to lay it out here.

**tl;dr: The `most-disruptive-allowed-action` support is in an API which Terraform will never call, and therefore this request cannot be implemented.**

To explain:  there are two types of update strategy that the Managed Instance Group (MIG) can use - `PROACTIVE` and `OPPORTUNISTIC`.  A `PROACTIVE` update will take place as soon as you update the MIG, and supports some strategies like maximum-unavailable-instances, create-before-destroy, that sort of thing.  The existing `minimal-action` support is for *proactive updates only*.

An `OPPORTUNISTIC` update is very different - since MIGs usually scale up and down, if an `OPPORTUNISTIC` update is requested, the MIG will scale down by removing instances with the old config, and scale up by adding instances with the new config, ensuring that instances will eventually be replaced.  It does not support any parameters for update strategy.

If you select `OPPORTUNISTIC`, then you can _also_ perform a manual update.  The manual update supports some parameters for update strategy, including minimal action and most-disruptive-allowed-action.  However, Terraform will never perform a manual update.  If you want Terraform to perform the update for you, you must select `PROACTIVE` updates.  `PROACTIVE` updates do _not_ include `most-disruptive-allowed-action`, though.

If you want a proactive update, and you also want the update to fail if it cannot be performed without recreating the instance (as I think you do), you should file an issue with GCP directly to support that use case.  When that is done, please let us know and we'll be happy to add Terraform support.
